### PR TITLE
ci: Upgrade compilers: GCC 13 and Clang 16

### PR DESCRIPTION
**Pull request by chfast**

### CODE REVIEW

File: .clang-tidy [hunk 1]
```diff
@@ -24,6 +24,7 @@ Checks: >
   clang-analyzer-*,
   cppcoreguidelines-*,
   -cppcoreguidelines-avoid-c-arrays,
+  -cppcoreguidelines-avoid-const-or-ref-data-members,
```
Comment (yperbasis) [inline .clang-tidy:27]: Why disable this check?

Comment (chfast) [reply .clang-tidy:27]: We have some structs with `const` data members. I can also suppress the warning locally.

Comment (gumb0) [reply .clang-tidy:27]: Maybe would be better to suppress locally, depending on how many there are such members. I couldn't find them yet.

Comment (chfast) [reply .clang-tidy:27]: There are 6 instances in this file (4 consts and 2 refs): https://github.com/ethereum/evmone/blob/master/lib/evmone/tracing.cpp

Comment (chfast) [reply .clang-tidy:27]: I think in this context all uses are actually good and neat, so I'm a bit skeptical about this warning.

**Review by yperbasis — APPROVED** [review]

File: .clang-tidy (hunk 1 above)
Comment (chfast) [reply .clang-tidy:27]: clang-tidy has recently added proper YAML format for these checks so we can later update the config, review the checks and at least add some comments why this is disabled.

Comment (chfast) [reply .clang-tidy:27]: For the reference, I noticed clang-tidy supports disabling a check in a block of code with `NOLINTBEGIN` / `NOLINTEND`. Yet this check is still not useful as it kills any kind of OOP. E.g. we have some valid uses of member as `evmc::VM& m_vm;`.

Comment (gumb0) [reply .clang-tidy:27]: Core guidelines' recommendation is to use a pointer in such cases (potentially wrapped in `not_null`)

**Pull request merged**